### PR TITLE
Add BM25 retriever

### PR DESCRIPTION
**Pull request by Ja-sonYun**
# Description
Add implementation of bm25 for document search.
It works well as keyword search, or using vector store and this together works better.

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [ ] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by logan-markewich — COMMENTED** [review]
Awesome, thanks for adding this! Just a few comments, but excited to merge this in!

I'm also inspired to add an advanced section to the notebook -- give me a few minutes and I'll have it pushed 😉

File: llama_index/retrievers/bm25_retriever.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    def __init__(
+        self,
+        index: VectorStoreIndex,
+        tokenizer: Callable[[str], list[str]],
+        similarity_top_k: int = DEFAULT_SIMILARITY_TOP_K,
+    ) -> None:
+        try:
+            from rank_bm25 import BM25Okapi
+        except ImportError:
+            raise ImportError("Please install rank_bm25 to use BM25Retriever")
```
Comment (logan-markewich) [inline llama_index/retrievers/bm25_retriever.py:24]: Can we modify the error here to include the needed `pip instal ...` command?

File: llama_index/retrievers/bm25_retriever.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        self._documents = cast(
+            list[Document], [doc for doc in self._docstore.docs.values()]
+        )
+        self._corpus = [self._tokenizer(doc.text) for doc in self._documents]
+
+        self.bm25 = BM25Okapi(self._corpus)
+
+    def _get_scored_nodes(self, query: str) -> list[NodeWithScore]:
+        tokenized_query = self._tokenizer(query)
+        doc_scores = self.bm25.get_scores(tokenized_query)
```
[the remark this answers is not in the document]
Comment (Ja-sonYun) [reply llama_index/retrievers/bm25_retriever.py:52]: Closer to 1 is more relevant. I added about this as a comment on the code.

**Comment by Ja-sonYun** [comment]
@logan-markewich  Should we add hybrid retriever as a class not only in notebook?

**Comment by logan-markewich** [comment]
@Ja-sonYun lets leave a dedicated component for another PR. In my mind, it's valuable to showcase how people can customize/build their own retrievers, rather than relying on pre-packaged solutions 💪🏻

**Pull request merged**